src/memoize.js

```diff
@@ -74,7 +74,7 @@ export function memoize(func, equalityCheck = eqCheck, orm) {
         // input arguments might result in a different set of
         // accessed models. On each run, we check if any new
         // models are accessed and add their invalidator functions.
-        session.accessedModels.forEach((modelName) => {
+        Object.keys(session.accessedModels).forEach((modelName) => {
```

Comment:
Currently this memoizer still just behaves the same way as it did before, only it examines the keys of `accessedModels` since it's not an Array anymore. This memoizer ignores the stored list of accessed instances.

The shape of the invalidator map needs updating to accommodate instances, and the actual comparison function needs to compare at the instance level instead of at the table level. I'm not sure what that would look like.

Reply:
hey, do you still need some help on this one, i can dig into it and propose some directions if you'd like.

Reply:
Yes! Let me know what you find <3

Reply:
something like this?

```js
export function eqCheck(a, b) {
  return a === b;
}

function instancesEqual(invalidatorMap, state) {
  return values(invalidatorMap).every(instanceEqualityChecker => instanceEqualityChecker(state));
}

export function memoize(func, equalityCheck = eqCheck, orm) {
  let lastOrmState = null;
  let lastResult = null;
  let lastArgs = null;
  const instanceEqualityCheckerByModelName = {};

  return (...args) => {
    const [dbState, ...otherArgs] = args;

    const argsAreEqual = lastArgs && otherArgs.every(
      (value, index) => equalityCheck(value, lastArgs[index]),
    );

    if (!argsAreEqual) {
      const dbIsEqual = lastOrmState === dbState ||
        instancesEqual(instanceEqualityCheckerByModelName, dbState);

      if (!dbIsEqual) {
        const session = orm.session(dbState);
        const newArgs = [session, ...otherArgs];
        const result = func(...newArgs);

        // If a selector has control flow branching, different
        // input arguments might result in a different set of
        // accessed models. On each run, we check if any new
        // models are accessed and add their invalidator functions.
        Object.keys(session.accessedModels).forEach((modelName) => {
          if (!instanceEqualityCheckerByModelName.hasOwnProperty(modelName)) {
            instanceEqualityCheckerByModelName[modelName] = nextState => {
              const lastModelState = lastOrmState[modelName];
              const nextModelState = nextState[modelName];

              if (lastModelState !== nextModelState) {
                return session.
                  accessedModels[modelName].
                  every(id => lastModelState.itemsById[id] === nextModelState.itemsById[id]);
              } else {
                return true;
              }
            };
          }
        });

        lastResult = result;
        lastOrmState = dbState;
        lastArgs = otherArgs;

      }
    }

    return lastResult;
  };
}

```

Reply:
If we save the model data instead of true to session.accessedModels, we could just make a === comparison?

Reply:
And move instanceEqualityCheckerByModelName to a function which takes the model name as an argument.

Reply:
I refactored your code a bit. Hope it's still functioning.. 

```javascript

import every from 'lodash/every';

const defaultEqualityCheck = (a, b) => a === b;

export function memoize(func, equalityCheck = defaultEqualityCheck, orm) {
  let lastOrmState = null;
  let lastResult = null;
  let lastArgs = null;

  const accessedModelInstancesAreEqual = (accessedModels, state) =>
    every(accessedModels, (accessedModelsForModelName, modelName) => {
      const nextModelState = state[modelName];

      if (lastOrmState === null) {
        return false;
      }

      const lastModelState = lastOrmState[modelName];
      if (lastModelState === state) {
        return true;
      }

      return every(
        accessedModelsForModelName,
        id => lastModelState.itemsById(id) === nextModelState.itemsById[id],
      );
    });

  const argsAreEqual = (a, b) =>
    a && b.every((value, index) => equalityCheck(value, lastArgs[index]));

  return (...args) => {
    const [nextOrmState, ...otherArgs] = args;

    if (argsAreEqual(lastArgs, otherArgs)) {
      return lastResult;
    }
    if (lastOrmState === nextOrmState) {
      return lastResult;
    }

    const session = orm.session(nextOrmState);
    if (accessedModelInstancesAreEqual(session.accessedModels, nextOrmState)) {
      return lastResult;
    }

    lastResult = func([session, ...otherArgs]);
    lastOrmState = nextOrmState;
    lastArgs = otherArgs;

    return lastResult;
  };
}
```

Reply:
I'm not a fan of multiple mid-function returns, and there were also some logic errors that I caught while refactoring. I think the logic is more direct on this one:

```js
import every from 'lodash/every';

const defaultEqualityCheck = (a, b) => a === b;

export function memoize(func, argEqualityCheck = defaultEqualityCheck, orm) {
  let lastOrmState = null;
  let lastResult = null;
  let lastArgs = null;

  // No need to check lastOrmState === nextOrmState again, because already checked that.
  const accessedModelInstancesAreEqual = (accessedModels, nextOrmState) => every(
    accessedModels,
    // Why is there two args for every here? I don't think lodash/every works with Objects...
    (accessedModelsForModelName, modelName) =>
      every(
        accessedModelsForModelName,
        id => lastOrmState[modelName].itemsById(id) === nextOrmState[modelName].itemsById[id],
      ),
  );

  return (...args) => {
    const [nextOrmState, ...nextArgs] = args;
    const session = orm.session(nextOrmState);

    if (lastOrmState) {
      if (
        (
          lastArgs &&
          lastOrmState === nextOrmState &&
          nextArgs.every((arg, i) => argEqualityCheck(arg, lastArgs[i]))
        ) || (
          accessedModelInstancesAreEqual(session.accessedModels, nextOrmState)
        )
      ) {
        return lastResult;
      }
    }

    lastResult = func(...[session, ...nextArgs]);
    lastOrmState = nextOrmState;
    lastArgs = nextArgs;

    return lastResult;
  };
}
```

Reply:
It was not my intent to offend you in any way! I'm really thankful for your contribution!
And: I never ran that code. I guess I should have made that clearer from the beginning :)

While .every only exists for Arrays in the ECMAScript Lodash implements this for Arrays and [Collections](https://lodash.com/docs/4.17.4#every):
```The predicate is invoked with three arguments: (value, index|key, collection).```

Sorry, I just had a hard time to understand what was going on so I tried to make it a bit clearer  for me and shared my result.

Early function returns are a big win for readability from my point of view, since you can immediately grasp what's the condition that lead to the return instead of having multiple conditions which exists of multiple sub-conditions themselves. I have a hard time resolving something like that in my head. Also, I guess the additional code generated while transpiling is negligible.

You were right, there were errors in the accessedModelInstancesAreEqual function!
```javascript
  const accessedModelInstancesAreEqual = (accessedModels, state) =>
    every(accessedModels, (accessedModelsForModelName, modelName) => {
      const nextModelState = state[modelName];

      if (lastOrmState === null) {
        return false;
      }

      const lastModelState = lastOrmState[modelName];
      if (lastModelState === nextModelState) {
        return true;
      }

      return 
        Object
          .keys(accessedModelsForModelName)
          .every(id => lastModelState.itemsById(id) === nextModelState.itemsById[id],
      );
    });
```

Reply:
@tiii - no offense taken. I'm just stating my opinion about mid-function returns. I'm ok with early function returns as well, so the accessedModelInstancesAreEqual is fine. It's more about mid-function returns or where the logic is not very clear, like in the main return function.

I actually thought there was an error on that because I saw it wasn't getting to regenerating the result, then I realized that there is actually and if/else condition happening. "If we are requesting with the same args and the accessed instances did not change, then use last result, otherwise recompute it."

To me it is very if/else-y if you will, but the code wasn't written that way, and it was confusing to reason the logic when I first looked at.

I respect everyone's different approaches so whoever is finalizing this PR can do what they like, they are both valid.

Reply:
Ah, I see! 👍  But I would argue that those are still early function returns if you only look at the returned function from memoize.

What do you think about checking the equality for all args not only otherArgs, that way we would respect a change to states as well?

```javascript
return (...args) => {
  const [nextOrmState, ...otherArgs] = args;

  if (argsAreEqual(lastArgs, args)) {
    return lastResult;
  }

  const session = orm.session(nextOrmState);
  if (accessedModelInstancesAreEqual(session.accessedModels, nextOrmState)) {
    return lastResult;
  }

  lastResult = func(...[session, ...otherArgs]);
  lastOrmState = nextOrmState;
  lastArgs = args;

  return lastResult;
};
```

Reply:
I just realized, that with ```instanceEqualityCheckerByModelName[modelName] = invalidatorFunction``` we basically introduced a global accessedModel store. Before that was kind of implicit since we did only memoize the tables.
An up to date version of memoize.js
```javascript
import every from 'lodash/every';
import merge from 'lodash/merge';

function defaultEqualityCheck(a, b) { a === b };

function const argsAreEqual(a, b) { b.every((value, index) => equalityCheck(value, a[index])) }

function accessedModelInstancesAreEqual(accessedModels, lastOrmState, nextOrmState) {
  every(accessedModels, (accessedModelsForModelName, modelName) => {
    const nextModelState = nextOrmState[modelName];

    if (lastOrmState === null) {
      return false;
    }

    const lastModelState = lastOrmState[modelName];
    if (lastModelState === nextOrmState) {
      return true;
    }

    return every(
      accessedModelsForModelName,
      id => lastModelState.itemsById(id) === nextModelState.itemsById[id],
    );
  });
}

export function memoize(func, equalityCheck = defaultEqualityCheck, orm) {
  let lastOrmState = null;
  let lastArgs = null;
  let lastResult = null;
  let allAccessedModels = {};

  return (...args) => {
    const [nextOrmState, ...otherArgs] = args;

    if (lastArgs && argsAreEqual(lastArgs, args)) {
      return lastResult;
    }

    if (accessedModelInstancesAreEqual(allAccessedModels, lastOrmState, nextOrmState)) {
      return lastResult;
    }

    const session = orm.session(nextOrmState);
    lastOrmState = nextOrmState;
    lastArgs = args;
    lastResult = func(...[session, ...otherArgs]);
    allAccessedModels = allAccessedModels.merge(session.accessedModels);

    return lastResult;
  };
}
```

Though we could persist that global accessedModel state in the memoize function I would rather persist it within the session itself? This makes the comment on the PR about the higher order component reseting accessedModels after render completion viable again..

@B-Reif is it ok if I hop into this with you, it get's kind of hard keeping track of all the "meta" thoughts on on this in the line comments... :P

Reply:
> What do you think about checking the equality for all args not only otherArgs, that way we would respect changes to the ormStates as well?

👍 

looks cleaner

Reply:
@tiii I added you to collab on my fork. Anyone else who'd like to commit there, feel free to reach out to me :>

Reply:
@B-Reif I'd love to be able to commit on your branch :slightly_smiling_face:

@tiii Caching all arguments breaks [the reduxIntegrationTest at this line](https://github.com/B-Reif/redux-orm/blob/feb596e520954768817c875a6d7ddebf6ff93469/src/test/testReduxIntegration.js#L57) because `db.getEmptyState()` creates a new object every time. No idea if this change in API is welcome or not, but I think the memoization should still work even when the root reference changes. We already compare accessed model instances, this should be enough.
 
`lastModelState.itemsById(id)` should probably be `lastModelState.itemsById[id]`.
`allAccessedModels.merge(session.accessedModels)` throws an error, this method does not exist. The object spread syntax may actually be enough here.

Further, I also very strongly favor early returns. We cannot split up the initial check into two if statements though, as this turns the boolean AND into a boolean OR. What do you think of this version? Also added some comments for clarity. Please correct me if I misunderstood something :grinning: 
```javascript
const defaultEqualityCheck = (a, b) => a === b;

const argsAreEqual = (a, b, equalityCheck) => (
    b.every((value, index) =>
        equalityCheck(value, a[index])
    )
);

const accessedModelInstancesAreEqual = (accessedModelInstances, lastOrmState, nextOrmState) => (
    every(accessedModelInstances, (accessedInstances, modelName) => {
        const nextModelState = nextOrmState[modelName];
        const lastModelState = lastOrmState[modelName];

        if (lastModelState === nextModelState) {
            return true;
        }

        const accessedIds = Object.keys(accessedInstances);
        return accessedIds.every(id =>
            lastModelState.itemsById[id] ===
            nextModelState.itemsById[id]
        );
    })
);

export function memoize(func, equalityCheck = defaultEqualityCheck, orm) {
    /* cached result of previous function call */
    let lastResult = null;
    /* cached arguments to previous function call (excluding ORM state) */
    let lastArgs = null;
    /* cached previous ORM state for the sake of comparing */
    let lastOrmState = null;
    /**
     * cached map of which model instances have been accessed
     * during previous function call (contains only IDs of accessed instances)
     * format (e.g.): { Book: { 1: true, 3: true } }
     */
    let lastAccessedModelInstances = {};

    return (...args) => {
        const [nextOrmState, ...otherArgs] = args;

        if (lastArgs &&
            lastOrmState &&
            argsAreEqual(lastArgs, otherArgs, equalityCheck) &&
            accessedModelInstancesAreEqual(lastAccessedModelInstances, lastOrmState, nextOrmState)) {
            /**
             * the instances that were accessed as well as
             * the arguments that were passed to func the previous time that
             * func was called have not changed
             */
            return lastResult;
        }

        /* previous result is no longer valid, update cached values */
        const session = orm.session(nextOrmState);
        lastOrmState = nextOrmState;
        lastArgs = otherArgs;
        lastResult = func(...[session, ...otherArgs]);

        lastAccessedModelInstances = {
            ...lastAccessedModelInstances,
            ...session.accessedModels,
        };

        return lastResult;
    };
}
```